### PR TITLE
fix(chunker): enforce strict chunk_token_num cap in naive_merge_docx

### PR DESCRIPTION
**Pull request by Nas01010101**
## Problem

`_merge_cks` (`rag/nlp/__init__.py`, used by `naive_merge_docx` — the "General" chunker for docx/doc/epub/json) decides whether to merge an incoming text unit into the running chunk by checking whether the *already-accumulated* total is over `chunk_token_num`, never the *projected* total (accumulated + incoming):

```python
if prev_text_ck < 0 or merged[prev_text_ck]["tk_nums"] >= chunk_token_num or has_custom:
    merged.append(cks[i]); ...  # start new chunk
    continue
merged[prev_text_ck]["tk_nums"] += cks[i]["tk_nums"]  # unconditional
```

This is the same bug class #17203 fixed for `naive_merge` / `RAGFlowTxtParser.parser_txt` (size check fired after the append instead of before it) — but #17203's "Out of scope" note states `naive_merge_docx` "already enforces the budget ... unchanged". That is not the case: `_merge_cks` has the same after-the-fact check.

Because the check only re-evaluates once per iteration rather than before every merge, the overshoot is not bounded to one extra unit — when text units are sized close to `chunk_token_num`, a chunk can grow to nearly double the budget before the next iteration notices.

Measured:

| Scenario | Budget | Result |
|---|--:|---|
| 20 paragraphs x 40 tokens | 128 | 5 chunks of 160 tokens (25% over) |
| 21 sections x 95 tokens | 100 | chunks of 190 tokens (90% over) |

This affects every `.docx` upload through the default "General" chunker (`rag/app/naive.py` is the only caller of `naive_merge_docx`).

## Fix

Mirror #17203's proactive-check shape: merge only if the projected total still fits the budget.

```python
incoming_tk = cks[i].get("tk_nums", 0)
if prev_text_ck < 0 or has_custom or merged[prev_text_ck]["tk_nums"] + incoming_tk > chunk_token_num:
    merged.append(cks[i]); ...
    continue
merged[prev_text_ck]["tk_nums"] = merged[prev_text_ck].get("tk_nums", 0) + incoming_tk
```

An already-oversized atomic unit (a single section with no internal delimiter, larger than the budget on its own) still cannot be sub-split by `_merge_cks` — it only merges, never splits — so it stays its own chunk, same as the current behaviour, and consistent with #17203 not attempting atom-level splitting at every call site either.

## Scope check on the Go rewrite

Not affected. `internal/ingestion/pipeline/template/ingestion_pipeline_general.json` routes docx (and doc/html/markdown/pdf) through a single shared `TokenChunker` component, which #17203 already patched (`internal/ingestion/component/chunker/token.go` was one of the files in that PR). The Go docx parsers (`internal/deepdoc/parser/docx`, `internal/parser/parser/docx_*.go`) only produce structure/IR and have no independent token-budget merge logic.

## Tests

New file `test/unit_test/rag/test_naive_merge_docx.py` — the first test coverage for `naive_merge_docx` / `_merge_cks`. 7 tests covering the projected-total check, packing efficiency for small units, the oversized-atomic-unit edge case, and the `has_custom` / image-passthrough branches.

Verified on Linux x86_64, Python 3.12:

```
# with this change reverted
FAILED test_naive_merge_docx_many_paragraphs_do_not_overshoot_budget - AssertionError: [160, 160, 160, 160, 160]
FAILED test_naive_merge_docx_near_budget_units_do_not_double_budget - AssertionError: [190, 190, 190, ...]
FAILED test_merge_cks_projected_total_check_no_overshoot - assert False
3 failed, 4 passed in 0.26s

# with this change
7 passed in 1.33s
```

`ruff check` / `ruff format --check` clean.

## Relates to

#17202, and #17203 — whose fix shape this mirrors and whose out-of-scope note this corrects.

### CODE REVIEW

**Comment by Nas01010101** [comment]
Merged main in. This branch was cut before #17808 and #18002 landed, and both of those touched `rag/nlp/__init__.py`, so it was worth checking whether they had already covered this. They haven't: `_merge_cks` on current main still reads

```python
if prev_text_ck < 0 or merged[prev_text_ck]["tk_nums"] >= chunk_token_num or has_custom:
```

which is still the accumulated-only check. The merge was clean and the diff against main is the same two files as before.

Re-ran on current main. `test_naive_merge.py`, `test_merge_paragraphs.py`, `test_delimiter_case_sensitive.py` and the new `test_naive_merge_docx.py` are 58 passed together. With `rag/nlp/__init__.py` restored to main's version, three of the new tests fail:

```
FAILED test_naive_merge_docx_many_paragraphs_do_not_overshoot_budget - AssertionError: [160, 160, 160, 160, 160]
FAILED test_naive_merge_docx_near_budget_units_do_not_double_budget - AssertionError: [190, 190, 190, ...]
FAILED test_merge_cks_projected_total_check_no_overshoot - assert False
```

so the 25% and 90% overshoots in the description still reproduce as written. `_merge_cks` has one caller, `naive_merge_docx`, so nothing else changes behaviour.

File: test/unit_test/rag/test_naive_merge_docx.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    # first two sections merge to 190 tokens (95 < 100 passes the stale
+    # check) before a third section trips the "already over budget" branch.
+    sentence = " ".join(["lorem"] * 95)
+    sections = [_section(sentence) for _ in range(21)]
+
+    chunks, _ = naive_merge_docx(sections, chunk_token_num=100)
+    texts = _nonempty_chunks(chunks)
+
+    assert all(c["tk_nums"] <= 100 for c in texts), [c["tk_nums"] for c in texts]
+
```
[the remark this answers is not in the document]
Comment (Nas01010101) [reply test/unit_test/rag/test_naive_merge_docx.py:115]: Good catch on the empty-list case, that assertion was vacuous. Added in a0bb257:

- `test_naive_merge_docx_near_budget_units_do_not_double_budget` now asserts `len(texts) > 1` and `sum(...) == 21 * 95`, matching what the many-paragraphs test above it already did.
- The image test asserts the text chunks still total 20 tokens, so the text either side of the image is accounted for.

Still 58 passed across the four `rag/nlp` suites, and reverting `rag/nlp/__init__.py` to main still fails the same three tests, at the `<= 100` line rather than vacuously.

**Comment by xugangqiang** [comment]
## Review of #17692 — `fix(chunker): enforce strict chunk_token_num cap in naive_merge_docx`

I read the diff, the PR description, all the CodeRabbit comments, and compared against the **current** `rag/nlp/__init__.py`. I also ran the merge logic to confirm behavior. **The code change itself is mechanically correct, but the PR's core premise is outdated, and merging it as-is would make the docx and txt chunker paths diverge in their cap semantics.**

### Key background (this drives the whole review)

The txt path `naive_merge` now defaults to `MergeStrategy.OVER_CAP` (`rag/nlp/__init__.py:1360`, contract docstring at `:1285` explicitly says "OVER_CAP has no hard cap"). I ran the two strategies on 21 units × 95 tokens, budget 100:

```
over_cap  -> chunks of 190 tokens (90% over budget)   # txt default today
under_cap -> chunks of 95  tokens (strict, never over) # what this PR makes docx
```

The **original** `naive_merge_docx` check at `rag/nlp/__init__.py:1765` (`merged[prev]["tk_nums"] >= chunk_token_num`) looks at the already-accumulated total — which is equivalent (up to the exact-budget boundary) to OVER_CAP's "close when accumulated > threshold". So **docx's current "overflow by one unit" behavior already matches txt's intended OVER_CAP contract.** It is not a docx-specific bug that #17203 fixed for txt.

This PR changes docx to `merged[prev]["tk_nums"] + incoming_tk > chunk_token_num` — that is exactly `UNDER_CAP` (strict) semantics. After this PR: **docx = strict cap, txt = soft cap (OVER_CAP).** The two paths drift apart.

### 1. Declared problem — does it actually solve it?

- Declared: "`_merge_cks` is the same bug class #17203 fixed for `naive_merge`; docx was left unpatched."
- Reality: `naive_merge` was later refactored by #17799 into the `OVER_CAP`/`UNDER_CAP` strategy enum, and its **default `OVER_CAP` intentionally allows a one-boundary overflow**. docx's current overflow matches txt's OVER_CAP by design — txt was *not* made stricter than docx. The "docx bug that txt doesn't have" premise does not hold under today's code.
- It *does* technically enforce a strict cap on docx (docx becomes UNDER_CAP). The problem is that this makes docx different from txt, not that the change "doesn't work."

### 2. Better approach

Do not silently switch docx alone to a strict cap. Two coherent options:

- **A (recommended if the current soft-cap contract stands):** Close the PR. `naive_merge_docx` already aligns with `naive_merge`'s OVER_CAP; nothing to fix.
- **B (only if strict caps are a deliberate system-wide decision):** This is a contract-level change and must flip **txt too** (to `UNDER_CAP`) and update the `merge_paragraphs` contract docstring. Fixing only docx is half the work.
- **Minimal-consistency fix if strict caps are wanted:** give `_merge_cks` a `MergeStrategy` parameter (default `OVER_CAP`, sharing `_merge_paragraph_groups`'s decision logic) instead of hand-writing a second, differently-shaped check. That keeps the two paths from drifting again.

### 3. Test coverage / regression risk

The new tests are good quality (boundary, packing efficiency, oversized atomic unit, `has_custom`, image) and use a monkeypatched word-count tokenizer for exact assertions — better than the existing sibling tests. But:

- **Contract risk:** `test_naive_merge_docx_near_budget_units_do_not_double_budget` asserts `tk_nums <= 100`, hard-coding "docx strictly never exceeds budget" as the expectation. If the correct decision is to keep OVER_CAP (option A), these assertions encode the *wrong* contract and conflict with txt's real behavior — they would lock docx into the inconsistent state.
- **Parity regression:** `naive_merge_docx` is the reference function for the live Go→Python parity harness. Changing docx chunk counts/sizes shifts every docx/doc/epub/json golden snapshot and requires a full re-baseline, otherwise parity tests fail wholesale.
- **User-side regression:** with the same `chunk_token_num=128`, docx would now emit smaller chunks than txt (fewer units per chunk), increasing chunk count → more embeddings / higher cost, and inconsistent experience across file types.

### 4. Does merging improve the system?

**No, as written.** It takes a path that was already consistent with txt and makes it stricter than txt. It breaks txt/docx/doc/epub/json chunk-behavior consistency, with no corresponding txt-side change. Unless the goal is global strict caps (then option B, fixing txt as well), this is net-negative.

### 5. Are the existing comments reasonable / how to fix them?

All four CodeRabbit comments stay inside the PR's (flawed) framing and **none of them catch the OVER_CAP/UNDER_CAP inconsistency**, which is the real issue.

1. "Assert text preservation in boundary tests (`all([])` vacuous pass)" (lines 100–112) — ✅ reasonable, already addressed (`len(texts) > 1` + `sum == 21*95`). A genuine catch.
2. "Add exact-budget case 60+40==100 still merges" (lines 146–155) — ✅ reasonable, low-cost; guards against a future `>`→`>=` regression. Worth adding, but only meaningful under option B.
3. "Remove unused `DEFAULT_DELIMITER`" (line 41) — ✅ reasonable (RUF001 + AGENTS.md "remove dead code"); low priority. CI does not run ruff, but house style favors deletion.
4. "Remove #17203 historical/compatibility wording" (lines 17–32) — ⚠️ partially reasonable but the direction needs correcting. AGENTS.md does forbid compatibility wording, but the more important fix is the **factual error**: the docstring claims "#17203's Out-of-scope note was incorrect." In fact, after #17799 txt also became OVER_CAP soft-cap, so the old "docx already enforces the budget, unchanged" note was *correct* — the PR's own understanding is outdated. Recommend deleting the misleading "#17203 was incorrect" claim and stating the current contract (docx should use the same `MergeStrategy` as `naive_merge`), rather than broadly "stripping historical narrative."

### Suggestion for the maintainer

Ask the author to first decide: **should docx be OVER_CAP or UNDER_CAP?** Before that is settled, the CodeRabbit test-level nitpicks are just polishing a PR that is pointing in the wrong direction. Don't accept them as a signal the PR is ready.

- If the soft-cap contract stands → close (option A).
- If strict caps are wanted → expand to a system-wide change that also flips txt (option B), or add a shared `MergeStrategy` parameter to `_merge_cks`.

**Comment by skbs-eng** [comment]
@Nas01010101 I just want to make sure that we also split oversized segments on whitespace with an exact token-count character-window binary search fallback, ensuring all chunks strictly fit the budget, just like in #17203 and RAGFlowHtmlParser._split_oversized_block.

**Comment by Nas01010101** [comment]
@skbs-eng Done in `ab952ab`.

`_split_oversized_unit` breaks an over-budget unit at whitespace-run ends; a whitespace-free run that is over budget on its own falls back to a character-window search inside that run. Both tiers share `_largest_fitting_cut`, which gallops to bracket the first overflow and then bisects inside it, so a probe tokenizes about one budget's worth of text rather than the whole remainder. It returns only an offset whose prefix a probe measured at `<= chunk_token_num`. The pieces rejoin to the input.

One difference from #17203 and `_split_oversized_block`: the count is taken on the assembled prefix instead of summed per atom. cl100k is not additive and whitespace atoms are not free — scoring them as zero lets a piece reach twice the cap on blank-line-heavy text (4 pieces, max 256 tokens at cap 128, against 7 pieces and max 128 here).

this is the docx path. `_merge_paragraph_groups`, `token_chunker._merge_text_chunks_by_token_size`, and the `_apply_overlap_unconditional` overlap can still exceed the cap. `_split_oversized_unit`, should i apply to this PR or a follow-up? @xugangqiang

**Comment by skbs-eng** [comment]
Can you implement the same fix for Go? Thanks.

**Comment by xugangqiang** [comment]
Could you clarify the intended contract for the docx path?

The Go production path (`mergeUnits` in `internal/ingestion/component/chunker/token.go`) already uses a hard-cap UNDER_CAP merge with oversized-unit expansion (`expandOversizedUnits`), and `token_merge_parity_test.go` documents this as a deliberate Go divergence from Python's OVER_CAP. So this PR, if merged, would align the Python docx path with the Go production behavior — that's a legitimate goal, but it should be stated explicitly.

Two questions so we can settle the direction before further review:

1. Is the intent to align the Python docx path with the Go side's hard-cap (UNDER_CAP + oversized-unit split) behavior? If yes, please say so in the PR description and align to it (the implementation already does; the framing needs to match).
2. If not, please state explicitly what the docx path's contract should be: stay OVER_CAP (soft cap, matching the current Python `naive_merge` default), UNDER_CAP (strict hard cap), or something else — and how that choice interacts with the txt path, which today defaults to OVER_CAP.

The code change itself is mechanically correct and the tests are solid; the blocker is just the unresolved system-level contract (see xugangqiang's review). Once the intended contract is stated explicitly, the path forward (align, close, or expand to a system-wide change) becomes clear.
